### PR TITLE
Publisher, Single, Completable onError* operators

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
ServiceTalk supports a recoverWith operator which allows for error
recovery [1]. However the signature of this method is more complex than
is often required for simple error recovery, and the name implies
"recovery" which is confusing if the return value is intended to
propagate an error.

[1] http://reactivex.io/documentation/operators/catch.html

Modifications:
- Add onError* methods to Publisher, Single, and Completable. The
  signatures are specialized for their respective operation (mapping
  Throwable, swallowing exceptions, recovering with a value) and also
  supports filtering by class type and custom Predicate. This naming
  convention is more commonly used by other JVM based rx implementations
  and is assumed to be more familiar to users.
- Deprecate recoverWith in favor of onErrorResume

Result:
Publisher, Single, and Completable have onError* operators which provide
more APIs tailored toward common error recovery scenarios.

### CODE REVIEW

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java [hunk 1]
```diff
@@ -402,16 +401,16 @@ public Completable closeAsyncGracefully() {
                         @Override
                         public Publisher<Resp> handle(final GrpcServiceContext ctx, final Publisher<Req> request) {
                             return request.firstOrError()
-                                    .recoverWith(t -> {
+                                    .onErrorMap(t -> {
                                         if (t instanceof NoSuchElementException) {
-                                            return failed(new GrpcStatus(INVALID_ARGUMENT, null,
+                                            return new GrpcStatus(INVALID_ARGUMENT, null,
```
Comment (bondolo) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java:406]: Worth it to initialize the cause?

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/OnErrorResumeCompletable.java [hunk 2]
```diff
@@ -89,7 +89,6 @@ public void onError(Throwable throwable) {
             try {
                 next = predicate.test(throwable) ? requireNonNull(nextFactory.apply(throwable)) : null;
             } catch (Throwable t) {
-                t.addSuppressed(throwable);
```
Comment (bondolo) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/OnErrorResumeCompletable.java:96]: Would the supressed be likely to have a different stack root? ie. from another thread with more context? It seems fairly low cost to keep it around unless the suppressed exception is truly redundant.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java (hunk 1 above)
Comment (Scottmitch) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java:406]: yah good point, not sure if there is a reason for it to be `null`. @idelpivnitskiy - wdyt?

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/OnErrorResumeCompletable.java (hunk 2 above)
Comment (Scottmitch) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/OnErrorResumeCompletable.java:96]: I will revert this change.

I think we need to do a more general investigation into `addSuppressed` usage which has the following potential issues:
- we don't use `addSuppressed` consistently
- it creates memory leaks if static/shared exceptions are thrown
- restrictions on usage such as self-add result in throwing which breaks async control flow

**Comment by Scottmitch** [comment]
build failure attributed to https://github.com/apple/servicetalk/issues/999

**Review by bondolo — APPROVED** [review]

**Comment by Scottmitch** [comment]
build failure attributed to https://github.com/apple/servicetalk/issues/1200

**Comment by Scottmitch** [comment]
build failure attributed to https://github.com/apple/servicetalk/issues/999

**Comment by Scottmitch** [comment]
another failure attributed to https://github.com/apple/servicetalk/issues/1200

**Pull request merged**